brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2018 The Brave Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "brave/utility/tor/tor_impl.h"
+
+#include <utility>
+
+#include "base/command_line.h"
+#include "base/process/kill.h"
+#include "base/process/launch.h"
+#include "base/single_thread_task_runner.h"
+#include "base/task_scheduler/post_task.h"
+#include "base/threading/thread.h"
+#include "base/threading/thread_task_runner_handle.h"
+
+namespace brave {
+
+TorLauncherImpl::TorLauncherImpl(
+    std::unique_ptr<service_manager::ServiceContextRef> service_ref)
+    : service_ref_(std::move(service_ref)) {}
+
+TorLauncherImpl::~TorLauncherImpl() {
+  if (tor_process_.IsValid()) {
+    tor_process_.Terminate(0, false);
+#if defined(OS_MACOSX)
+    base::PostTaskWithTraits(
+        FROM_HERE, {base::MayBlock(), base::TaskPriority::BACKGROUND},
+        base::BindOnce(&base::EnsureProcessTerminated, Passed(&tor_process_)));
+#else
+    base::EnsureProcessTerminated(std::move(tor_process_));
+#endif
+  }
+}
+
+#if !defined(OS_WINDOWS)
+void SIGCHLDHandler(int signal) {
+  pid_t p;
+  int status;
+
+  while ((p = waitpid(-1, &status, WNOHANG)) != -1) {
+    LOG(ERROR) << "Tor process terminated";
+    // TODO(darkdh): handle tor got killed by other factors
+  }
+}
+#endif
+
+void TorLauncherImpl::Launch(const base::FilePath& tor_bin,
+                             const std::vector<std::string>& args,
+                             LaunchCallback callback) {
+  if (tor_process_.IsValid())
+    return;
+  base::CommandLine cmdline(tor_bin);
+  for (size_t i = 0; i < args.size(); ++i) {
+    cmdline.AppendArg(args[i]);
+  }
+  base::LaunchOptions launchopts;
+#if defined(OS_LINUX)
+  launchopts.kill_on_parent_death = true;
+#endif
+  tor_process_ = base::LaunchProcess(cmdline, launchopts);
+
+  bool result;
+  if (tor_process_.IsValid())
+    result = true;
+  else
+    result = false;
+
+  if (callback)
+    std::move(callback).Run(result);
+
+  child_monitor_thread_.reset(new base::Thread("child_monitor_thread"));
+  if (!child_monitor_thread_->Start()) {
+    NOTREACHED();
+  }
+
+  child_monitor_thread_->task_runner()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&TorLauncherImpl::MonitorChild, base::Unretained(this)));
+}
+
+void TorLauncherImpl::IsTorAlive(IsTorAliveCallback callback) {
+  if (callback)
+    std::move(callback).Run(tor_process_.IsValid());
+}
+
+void TorLauncherImpl::MonitorChild() {
+#if !defined(OS_WINDOWS)
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+#else
+  WaitForSingleObject(tor_process_.Handle(), INFINITE);
+  LOG(ERROR) << "Tor process terminated";
+  // TODO(darkdh): handle tor got killed by other factors
+#endif
```

Comment:
This seems to have two different actions:

1. On non-Windows, the thread installs a signal handler and exits.
2. On Windows, the thread waits for the child process to finish.

Did you mean to call `pause` or `sigsuspend` in the non-Windows case so that the thread also waits, or is that not necessary with the Chromium thread concept -- i.e., do Chromium threads live forever handling signals as long as you don't explicitly terminate them?

(Also, tiny nit: I suggest using OS_POSIX for case (1), and OS_WINDOWS for case (2), and maybe adding an `#else / #error` so that attempts to build for non-POSIX/Windows fail more obviously.)

Reply:
There will be a `UtilityMain` executing `RunLoop` for this service, you can treat it as living forever until browser process shutdown or explicitly kill this process.